### ISSUE TITLE
Referencing unrelated documents in sources

### ISSUE THREAD

**Issue by RohanKaran** [h2oai/h2ogpt#323]
In the response it is referencing some documents which has no relation with the question asked with a random score assigning with it. Even if it's answering from llm data source (not from UserData) it's still referencing those documents. Can you clarify how and why it is referencing those documents from UserData which has no similarity from the question asked and how is the score assigned to those documents?

**Comment by pseudotensor** [comment]
Hi.  I also feel like the scores are a bit high for unrelated things.  I didn't find any bugs, but I'll check on it again.  Thanks!

**Comment by pseudotensor** [comment]
Hi @RohanKaran I checked both FAISS and chroma in debug mode and both give the same score for the documents in various checks.

However, I find that the instruct embedding gives relatively flat score distribution, making it hard to tell relevant from irrelevant.  Something like `--hf_embedding_model=sentence-transformers/all-MiniLM-L6-v2` gives a distribution that's sharper and seems to distinguish better.

**Comment by RohanKaran** [comment]
> Hi @RohanKaran I checked both FAISS and chroma in debug mode and both give the same score for the documents in various checks.
> 
> However, I find that the instruct embedding gives relatively flat score distribution, making it hard to tell relevant from irrelevant. Something like `--hf_embedding_model=sentence-transformers/all-MiniLM-L6-v2` gives a distribution that's sharper and seems to distinguish better.

That helped a bit, now the scores are not absolutely random. 
But in some cases if I normally say "hi" or any normal conversation having All_Relevant selected as document choice, it gives some weird answers and quotes a document with 0 or nearly 0 scores. Is there a way to not use UserData and only LLM db in case the scores are low.

**Comment by pseudotensor** [comment]
You can check the cut_distance to be lower than 1.64 here:
https://github.com/h2oai/h2ogpt/blob/2c0b4120069cc9892703492586b8902c209ba20b/generate.py#L1546

Maybe try 1.5.  It might lead to cases when should have used document but didn't.

Also, the 0 score is because I show the score like this: https://github.com/h2oai/h2ogpt/blob/2c0b4120069cc9892703492586b8902c209ba20b/gpt_langchain.py#L2191-L2192

So I take 1.5 - distance divided by 1.5.  So a score of 0 will show up when the "max" triggers as occurs when "score" is > 1.5 but still < 1.64.

So UI/CLI scores of 0 will no longer show if you use `cut_distanct=1.5` and very nearly 0 scores will be removed if you make it even a bit smaller.

**Comment by RohanKaran** [comment]
> You can check the cut_distance to be lower than 1.64 here:
> 
> https://github.com/h2oai/h2ogpt/blob/2c0b4120069cc9892703492586b8902c209ba20b/generate.py#L1546
> 
> Maybe try 1.5. It might lead to cases when should have used document but didn't.
> 
> Also, the 0 score is because I show the score like this:
> 
> https://github.com/h2oai/h2ogpt/blob/2c0b4120069cc9892703492586b8902c209ba20b/gpt_langchain.py#L2191-L2192
> 
> So I take 1.5 - distance divided by 1.5. So a score of 0 will show up when the "max" triggers as occurs when "score" is > 1.5 but still < 1.64.
> 
> So UI/CLI scores of 0 will no longer show if you use `cut_distanct=1.5` and very nearly 0 scores will be removed if you make it even a bit smaller.

Thanks, solves the problem.

**Comment by RohanKaran** [comment]
Hi @pseudotensor  why are the reference scores same for two different answers from different models? Does it measure the relevance between doc sources and the prompt, not doc sources and response?
![image](https://github.com/h2oai/h2ogpt/assets/38720754/fb31e5cc-73a4-472c-b7f8-9456512a9059)

**Comment by pseudotensor** [comment]
Correct, the score is currently related to the distance between the query and chunk of data, not the response and chunk of data.  The latter can be done though.

**Comment by RohanKaran** [comment]
> Correct, the score is currently related to the distance between the query and chunk of data, not the response and chunk of data. The latter can be done though.

The latter should be better, because we have to assume that LLMs will give best answers. By measuring cosine distance with the response, we should get the correct chunks. The ranking of chunks also will be good in that case.
